bazel/closure_grpc_web_library.bzl

```diff
@@ -0,0 +1,195 @@
+  ])
+
+  files = []
+  for src in sources:
+    name = '{}.grpc.js'.format(
+        '.'.join(src.path.split('/')[-1].split('.')[:-1]))
+    js = actions.declare_file(name)
+    files.append(js)
+
+    args = proto_include_paths + [
+        '--plugin=protoc-gen-grpc-web={}'.format(protoc_gen_grpc_web.path),
+        '--grpc-web_out={options},out={out_file}:{path}'.format(
+            options = grpc_web_out_common_options,
+            out_file = name,
+            path = js.path[:js.path.rfind('/')],
+        ),
+        src.path
+    ]
+
+    actions.run(
+        inputs = [protoc_gen_grpc_web] + all_sources,
+        outputs = [js],
+        executable = protoc,
+        arguments = args,
+        progress_message =
+            _generate_closure_grpc_web_src_progress_message(name),
+    )
+
+  return files
+
+_error_multiple_deps = ''.join([
+  "'deps' attribute must contain exactly one label ",
+  "(we didn't name it 'dep' for consistency). ",
+  "We may revisit this restriction later.",
+])
+
+def _closure_grpc_web_library_impl(ctx):
+  if len(ctx.attr.deps) > 1:
+    # TODO(yannic): Revisit this restriction.
+    fail(_error_multiple_deps, 'deps');
+
+  dep = ctx.attr.deps[0]
+
+  srcs = _generate_closure_grpc_web_srcs(
+      actions = ctx.actions,
+      protoc = ctx.executable._protoc,
+      protoc_gen_grpc_web = ctx.executable._protoc_gen_grpc_web,
+      import_style = 'closure',
+      mode = 'grpcwebtext',
+      sources = dep.proto.direct_sources,
+      transitive_sources = dep.proto.transitive_imports,
+  )
+
+  deps = unfurl(ctx.attr.deps, provider = "closure_js_library")
+  deps += [
+      ctx.attr._grpc_web_abstractclientbase,
+      ctx.attr._grpc_web_clientreadablestream,
+      ctx.attr._grpc_web_error,
+      ctx.attr._grpc_web_grpcwebclientbase
+  ]
+
+  suppress = [
+      "misplacedTypeAnnotation",
+  ]
+
+  library = closure_js_library_impl(
+      actions = ctx.actions,
+      label = ctx.label,
+      workspace_name = ctx.workspace_name,
+
+      srcs = srcs,
+      deps = deps,
+      testonly = ctx.attr.testonly,
+      suppress = suppress,
+      lenient = False,
+
+      closure_library_base = ctx.files._closure_library_base,
+      _ClosureWorker = ctx.executable._ClosureWorker,
+  )
+  return struct(
+      exports = library.exports,
+      closure_js_library = library.closure_js_library,
+      # The usual suspects are exported as runfiles, in addition to raw source.
+      runfiles = ctx.runfiles(files = srcs),
+  )
+
+closure_grpc_web_library = rule(
+    implementation = _closure_grpc_web_library_impl,
+    attrs = {
+        'deps': attr.label_list(
+            mandatory = True,
+            providers = ['proto', 'closure_js_library'],
+            # The files generated by this aspect are required dependencies.
+            aspects = [closure_proto_aspect],
+        ),
+        'import_style': attr.string(
+            default = 'closure',
+            values = ['closure', 'commonjs'],
```

Comment:
I think I'd like to remove `commonjs` as a possible value for now. Most of `rules_closure` doesn't work with `commonjs` anyway.

Reply:
Is it because the output of this rule is expected to mostly feed into other `closure_js_library` rule, which expects `goog.require()` style includes anyways?

Reply:
The restriction to `goog.require()` is required because `closure_proto_{aspect,library}` (which generates the protobuf messages for us to avoid duplicate provide errors) does not support `commonjs` import style (yet). I will create a follow-up PR once this is fixed in rules_closure (Maybe we can create an issue for this and assign it to me).